File: libraries/Runtime/BridJ/src/test/velocity/com/bridj/CallTest.cpp

#foreach ($prim in $primitivesNoBool)

TEST_API j${prim.Name} test_incr_${prim.Name}(j${prim.Name} value) {
	return (j${prim.Name})(value + 1);
}

TEST_API j${prim.Name} test_callback_${prim.Name}_${prim.Name}(j${prim.Name} (*cb)(j${prim.Name}), j${prim.Name} value) {
	j${prim.Name} ret = cb(value);
	return ret;
}


#foreach ($n in [9..9])

TEST_API j${prim.Name} test_add${n}_${prim.Name}(#foreach ($i in [1..$n])#if($i > 1), #end j${prim.Name} arg$i#end) {
	j${prim.Name} tot = (j${prim.Name})0;
	j${prim.Name} fact = (j${prim.Name})1;
#foreach ($i in [1..$n])
	j${prim.Name} v$i = (j${prim.Name})(fact * ($i + 1));
	fact *= (j${prim.Name})2;
	tot += v$i;
#end
	return tot;
}
#end

#end

File: libraries/Runtime/BridJ/src/test/velocity/com/bridj/CallTest.java
package com.bridj;

import com.bridj.ann.*;
import static com.bridj.Functional.*;
import java.util.*;
import org.junit.*;
import static org.junit.Assert.*;

@Library("test")
@com.bridj.ann.Runtime(CRuntime.class)
public class CallTest {

	public CallTest() {
		BridJ.register(CallTest.class);
	}
	
#foreach ($prim in $primitivesNoBool)

	/// Returns value + 1
	public static native ${prim.Name} test_incr_${prim.Name}(${prim.Name} value);

	@Test
	public void testIncrement${prim.CapName}() {
		for (${prim.Name} value : new ${prim.Name}[] { (${prim.Name})0, (${prim.Name})1, (${prim.Name})-1 }) {
			${prim.Name} ret = test_incr_${prim.Name}(value);
			${prim.Name} incr = (${prim.Name})(value + 1);
			assertEquals(incr, ret#if($prim.Name == "float" || $prim.Name == "double"), 0#end);
		}
	}
	
	/*
	/// Returns cb.apply(value)
	public static native ${prim.Name} test_callback_${prim.Name}_${prim.Name}(Func1<${prim.WrapperName}, ${prim.WrapperName}> cb, ${prim.Name} value);
	
	@Test
	public void testCallback_${prim.Name}() {
		Func1<${prim.WrapperName}, ${prim.WrapperName}> cb = new Func1<${prim.WrapperName}, ${prim.WrapperName}>() {
			public ${prim.WrapperName} apply(${prim.WrapperName} value) {
				return (${prim.Name})(value + 1);
			}
		};
		for (${prim.Name} value : new ${prim.Name}[] { (${prim.Name})0, (${prim.Name})1, (${prim.Name})-1 }) {
			${prim.Name} ret = test_callback_${prim.Name}_${prim.Name}(cb, value);
			${prim.Name} incr = (${prim.Name})(value + 1);
			assertEquals(incr, ret#if($prim.Name == "float" || $prim.Name == "double"), 0#end);
		}
	}
	*/

	
#foreach ($n in [9..9])
	public static native ${prim.Name} test_add${n}_${prim.Name}(#foreach ($i in [1..$n])#if($i > 1), #end${prim.Name} arg$i#end);
	
	@Test
	public void testAdd${n}${prim.CapName}() {
		${prim.Name} expectedTot = (${prim.Name})0;
		${prim.Name} fact = (${prim.Name})1;
#foreach ($i in [1..$n])
		${prim.Name} arg$i = (${prim.Name})(fact * ($i + 1));
		fact *= (${prim.Name})2;
		expectedTot += arg$i;
#end
		${prim.Name} tot = test_add${n}_${prim.Name}(#foreach ($i in [1..$n])#if($i > 1),#end arg$i#end);
		assertEquals(expectedTot, tot#if($prim.Name == "float" || $prim.Name == "double"), 0#end);
	}
#end


#end
}
